modules/packages/Sort.chpl

```diff
@@ -1335,13 +1335,13 @@ module ShellSort {
     // Analysis of Shellsort and Related Algorithms 1996
     // and see Marcin Ciura - Best Increments for the Average Case of Shellsort
     // for the choice of these increments.
-    var n = 1 + end - start;
     var js,hs:idxType;
     var v,tmp:Data.eltType;
     const incs = (701, 301, 132, 57, 23, 10, 4, 1):(8*idxType);
     for h in incs {
-      // skip past cases in which the 'incs' value is too big for idxType
-      if h < 0 || h:uint > max(idxType):uint then
+      // skip past cases in which the 'incs' value was too big for
+      // idxType, or in which h+start will overflow idxType.
+      if h < 0 || h >= max(idxType):uint - start:uint then
```

Comment:
@cassella: This seems problematic if `start` is a negative integer due to casting it to `uint`?  Or are you relying on negative int -> uint wraparound properties that I usually just shy away from thinking about?

Reply:
@bradcray: The latter.  I thought it was right when I wrote it, and I still think it's right, but I haven't been able to summon up the brainpower tonight to be able to explain it coherently.

Reply:
I'm having a hard time putting it into words, but I haven't changed my mind.

It may help to consider that there aren't distinct signed and unsigned subtract instructions.  The same bit value results whichever way you're interpreting the operands.  So you can interpret them whichever way helps, whichever way they're declared.

If `start` is negative, you can think of the subtraction in `int(64)`.  It will underflow if `idxType` was `int(64)` to start with, but the result will be the `uint(64)` describing the difference between them.


It may just be a matter of being confident that the right thing happens when `start` is cast to `uint(64)`, and that the subtraction works the right way.

Reply:
@cassella: What do you think about the following variation as a further step in the right direction:

```chapel
  const incs = (701, 301, 132, 57, 23, 10, 4, 1);

  for hh in incs {
    // skip past cases in which the 'incs' value was too big for              
    // idxType, or in which h+start will overflow idxType.                    
    if hh < max(idxType):uint - start:uint {
      const h = hh:idxType;
      // do the normal computation using `h` here
    }
  }
```

Motivations:

* fixes an issue in my previous approach that I'd missed in which 301 happens to wrap around for int(8) to 45, so doesn't get filtered out by my vague/confusing `h < 0` test as intended
* related: avoids the lame-ish/confusing-ish downcast of `incs` to `8*idxType`
* avoids relying on a `continue` for simpler control flow
* still passes your test

I think I've convinced myself about the "cast to uint and subtract" approach, where my explanation to my future self would be:

* we'd ideally like to compare `start + hh` < `max(idxType)`
* however, `start + hh` could wrap around, so we move `hh` to the other side:  `start < max(idxType) - hh`
* however, for `idxType = int`, the difference between two ints can exceed max(int); so to ensure the result doesn't wrap around, we do the math in terms of uints:  `start < max(idxType):uint - hh:uint`
* I always forget this, but: `start` need not be cast to `uint` to make this operator work because Chapel supports `int < uint` comparisons by first checking for `int < 0`, returning true if it is; casting to uint and doing the comparison otherwise

Reply:
`hh` can still problematically exceed `max(idxType)`.  E.g., the `132` for an `int(8)`-based range when `start` is say `-50`.  `132 < 127:uint - -50 == 177`, but `132:int(8) < 0`.

(The problem being that `max(idxType):uint - start:uint > max(idxType):uint`, so `hh` being less than the former doesn't imply it's less than the latter.)

Maybe that indicates I should add more test cases.

I do like getting rid of the tuple-cast.

When the control flow is

```
loop {
  if (usual case) {
    do the normal case at another level of indentation;
  }
}
```

and the normal case is more than a few lines, by and large I prefer

```
loop {
  if (exceptional case)
    continue;
  do normal case
}
```

That may be partly influenced by Linux kernel programming where 8-space indentation is the rule.  But I also have an easier time keeping track in my head if in the normal case code we're completely done with the exceptional case and I don't have to keep in mind that once the `if()` block ends the exceptional case will rejoin.  Especially if the end of the `if()` block doesn't fit on the same screen as its beginning.

I'll see how the loop looks with some of these changes, and add more testing.

Reply:
I'm having a bit of trouble with that test, due to #17264.

Reply:
Even apart from that issue, I can't work out a case where a negative `h` that doesn't have `h+start` overflow `max(idxType)` causes a problem.  The constraints put on `hh` and `hs`, the iteration over `hs..end`, and the calculations being done in `idxType` seem to conspire to keep the array accesses within the bounds of `start..end`.

It's not a deliberate part of the algorithm, so I'm inclined to have it skip `hh` that exceed `idxType`, but I can't think of a way to write a test that will fail without it.

Reply:
I updated to use the new form.

I still can't find a test that gets an out-of-bounds condition due to `hh` exceeding `max(idxType)`, but executing the iteration with an effectively random choice of the shell size (I think) seems counter to the algorithm, so I am still skipping those values of `hh`.

I squashed the commit since it changes nearly every changed line.  I think every line of the commit header is still accurate as written, though I did also add a note about dropping the unused `n` and that one of the test cases now includes `start < 0`.

The other change was to have the test iterate over `0..#loStartHi.size` instead of hardcoding `3`.

Reply:
Thanks for catching the bug in my thinking, Paul.  I'm giving this one final test, out of an abundance of caution, then plan to merge.